packages/outbox-core/lib/outbox.ts

```diff
@@ -0,0 +1,170 @@
+ * - CREATED - entry was created and is waiting to be processed to publish actual event
+ * - ACKED - entry was picked up by outbox job and is being processed
+ * - SUCCESS - entry was successfully processed, event was published
+ * - FAILED - entry processing failed, it will be retried
+ */
+export type OutboxEntryStatus = 'CREATED' | 'ACKED' | 'SUCCESS' | 'FAILED'
+
+export type OutboxEntry<SupportedEvent extends CommonEventDefinition> = {
+  id: string
+  event: SupportedEvent
+  data: Omit<CommonEventDefinitionPublisherSchemaType<SupportedEvent>, 'type'>
+  precedingMessageMetadata?: Partial<ConsumerMessageMetadataType>
+  status: OutboxEntryStatus
+  created: Date
+  updated?: Date
+  retryCount: number
+  lockedUntil?: Date
+}
+
+/**
+ * Takes care of persisting and retrieving outbox entries.
+ *
+ * Implementation is required:
+ * - in order to fulfill at least once delivery guarantee, persisting entries should be performed inside isolated transaction
+ * - to return entries in the order they were created (UUID7 is used to create entries in OutboxEventEmitter)
+ * - returned entries should not include the ones with 'SUCCESS' status
+ */
+export interface OutboxStorage<SupportedEvents extends CommonEventDefinition[]> {
+  create(
+    outboxEntry: OutboxEntry<SupportedEvents[number]>,
+  ): Promise<OutboxEntry<SupportedEvents[number]>>
+
+  update(
+    outboxEntry: OutboxEntry<SupportedEvents[number]>,
+  ): Promise<OutboxEntry<SupportedEvents[number]>>
+
+  /**
+   * Returns entries in the order they were created. It doesn't return entries with 'SUCCESS' status. It doesn't return entries that have been retried more than maxRetryCount times.
+   *
+   * For example if entry retryCount is 1 and maxRetryCount is 1, entry MUST be returned. If it fails again then retry count is 2, in that case entry MUST NOT be returned.
+   */
+  getEntries(maxRetryCount: number): Promise<OutboxEntry<SupportedEvents[number]>[]>
+}
+
+/**
+ * Main logic for handling outbox entries.
+ *
+ * If entry is rejected, it is NOT going to be handled during the same execution. Next execution will pick it up.
+ */
+export class OutboxProcessor<SupportedEvents extends CommonEventDefinition[]> {
+  constructor(
+    private readonly outboxStorage: OutboxStorage<SupportedEvents>,
+    private readonly eventEmitter: DomainEventEmitter<SupportedEvents>,
+    private readonly maxRetryCount: number,
+  ) {}
+
+  public async processOutboxEntries(context: JobExecutionContext) {
+    const entries = await this.outboxStorage.getEntries(this.maxRetryCount)
+
+    for (const entry of entries) {
+      try {
+        const updatedEntry = await this.outboxStorage.update({
+          ...entry,
+          updated: new Date(),
+          status: 'ACKED',
+        })
+
+        await this.eventEmitter.emit(entry.event, entry.data, entry.precedingMessageMetadata)
+
+        await this.outboxStorage.update({ ...updatedEntry, updated: new Date(), status: 'SUCCESS' })
```

Comment:
that's a lot of update operations in a loop! in CockroachDB that would absolutely trash the perf, but even PostgreSQL is not going to appreciate it

wonder if a better approach would be to accumulate individual ack/emission facts in redis first, and then do a batch update. and when a job starts, first check if there are any leftovers from the previous attempt

Reply:
I think that's fine. In my opinion, it's an implementation detail.

E.g. `outbox-prisma-adapter` could accumulate such entries in memory and hit `updateMany` every 20-30 entries.
```js
update(entry) {
  this.entries.push(entry)

  if (this.entries.length > 20) {
    await this.updateMany(this.entries)
    this.entries = []
  }
}
```
The only thing I need to add is to extend `OutboxStorage` with method `flush()` that would be triggered after we finish processing. It could signal to implementation, that there won't be more entries, so whatever is left in `this.entries` could be persisted.

Interface itself is open, also for Redis wrapper for prisma/drizzle adapters. It could be construced in following way:
```js
new OutboxPeriodicJob(
  new RedisOutboxAccumulator(
    diContainer.cradle.redis, //Also we'd need some extra params to configure behavior of accumulator
    new PrismaOutboxAdapter(diContainer.cradle.prisma, 'outbox_entry'), //We provide any impl of OutboxStorage here
  ),  
  eventEmitter,
  dependencies,
  maxRetryCount
  ONE_SEC 
)
```

Reply:
> E.g. outbox-prisma-adapter could accumulate such entries in memory and hit updateMany every 20-30 entries.

How would you avoid double publish in case service crashes?

Reply:
> How would you avoid double publish in case service crashes?

Also discussed offline, consumers need to take care of it. I will describe everything later in README.md.

Reply:
Done, provided in memory accumulator.

Reply:
Let me also update storage method to have `updateMany`, `updateFailures`.